File: program_stack/algorithms/graphs/topological_sort/cpp/topological_sort.cpp
#include <algorithm>
#include <cstdlib>
#include <iostream>
#include <optional>
#include <queue>
#include <vector>

// Kahn's algorithm. nullopt if the graph has a cycle.
std::optional<std::vector<int>> topologicalSort(const std::vector<std::vector<int>>& adj) {
    int n = static_cast<int>(adj.size());
    std::vector<int> indeg(n, 0);
    for (int u = 0; u < n; ++u)
        for (int v : adj[u]) ++indeg[v];
    std::queue<int> q;
    for (int i = 0; i < n; ++i)
        if (indeg[i] == 0) q.push(i);
    std::vector<int> order;
    while (!q.empty()) {
        int u = q.front();
        q.pop();
        order.push_back(u);
        for (int v : adj[u]) {
            if (--indeg[v] == 0) q.push(v);
        }
    }
    if (static_cast<int>(order.size()) != n) return std::nullopt;
    return order;
}

int main() {
    std::vector<std::vector<int>> adj = {{1, 2}, {3}, {3}, {}};
    auto order = topologicalSort(adj);
    std::vector<int> expect = {0, 1, 2, 3};
    if (!order || *order != expect) {
        std::cerr << "unexpected order\n";
        return 1;
    }
    adj = {{1}, {0}};
    order = topologicalSort(adj);
    if (order) {
        std::cerr << "expected cycle detection failure\n";
        return 1;
    }
    std::cout << "topological_sort: ok\n";
    return 0;
}

File: program_stack/algorithms/graphs/topological_sort/java/TopologicalSort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class TopologicalSort {
    /**
     * Kahn's algorithm. Returns a valid order, or null if the graph has a cycle.
     */
    public static List<Integer> topologicalSort(List<List<Integer>> adj) {
        int n = adj.size();
        int[] indeg = new int[n];
        for (int u = 0; u < n; u++) {
            for (int v : adj.get(u)) indeg[v]++;
        }
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < n; i++) if (indeg[i] == 0) q.add(i);
        List<Integer> order = new ArrayList<>();
        while (!q.isEmpty()) {
            int u = q.poll();
            order.add(u);
            for (int v : adj.get(u)) {
                if (--indeg[v] == 0) q.add(v);
            }
        }
        return order.size() == n ? order : null;
    }

    private static List<List<Integer>> graph(int[][] edges, int n) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for (int[] e : edges) adj.get(e[0]).add(e[1]);
        return adj;
    }

    public static void main(String[] args) {
        // 0→1, 0→2, 1→3, 2→3  → [0,1,2,3]
        List<Integer> order = topologicalSort(graph(new int[][]{{0,1},{0,2},{1,3},{2,3}}, 4));
        if (order == null || !order.equals(Arrays.asList(0, 1, 2, 3))) {
            System.err.println("unexpected order: " + order);
            System.exit(1);
        }
        // cycle 0→1→0
        order = topologicalSort(graph(new int[][]{{0,1},{1,0}}, 2));
        if (order != null) {
            System.err.println("expected cycle detection failure");
            System.exit(1);
        }
        System.out.println("topological_sort: ok");
    }
}
